### PR TITLE
[GH-500]: Fixed github issue "Emoji synchronisation for PR/issue on github"

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**
**Summary** 

- When the GitHub bot creates a post, it's always related to some specific issue/PR or comment. This PR has added the functionality that if a user reacts to the post in Mattermost with one of the 8 emojis supported for reactions on GitHub, the plugin will proxy this reaction to the appropriate PR/comment on GitHub. Likewise, if the user removes their emoji reaction in Mattermost, it will remove it in GitHub. The supported event notifications are:- 
- Issue/PR creations
- Issue/PR comments
- PR review comments

- This PR is an extension of [PR #575](https://github.com/mattermost/mattermost-plugin-github/pull/575)

- [Issue #500](https://github.com/mattermost/mattermost-plugin-github/issues/500)

### CODE REVIEW

File: server/plugin/plugin.go [hunk 1]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
...
+
+	info, appErr := p.getGitHubUserInfo(reaction.UserId)
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogDebug("Error in getting user info", "error", appErr.Message)
+		}
+		return
+	}
+
+	ghClient := p.githubConnectUser(context.Background(), info)
```
Comment (javaguirre) [inline server/plugin/plugin.go:327]: I'd rather have this last part of the code in another function/method so it's easier to follow and unittest.

File: server/plugin/plugin.go [hunk 2]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
...
+		}
+	default:
+		return
+	}
+}
+
+func (p *Plugin) ReactionHasBeenRemoved(c *plugin.Context, reaction *model.Reaction) {
+	githubEmoji := p.emojiMap[reaction.EmojiName]
+	if githubEmoji == "" {
+		return
```
Comment (javaguirre) [inline server/plugin/plugin.go:353]: logging info or warning here to know why the emoji is not in the list (with reaction.EmojiName to show on the log)?

File: server/plugin/plugin.go [hunk 3]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogDebug("Error in getting user info", "error", appErr.Message)
+		}
+		return
+	}
+
+	ghClient := p.githubConnectUser(context.Background(), info)
+	owner, repo := orgRepo[0], orgRepo[1]
+	switch objectType {
```
Comment (javaguirre) [inline server/plugin/plugin.go:371]: This is more or less the same switch as above, so we could have a function selecting the case and two methods deciding what to return depending of what we want right? Unitesting will also be easier that way.

**Review by javaguirre — CHANGES_REQUESTED** [review]

**Comment by javaguirre** [comment]
Thank you for the work @Kshitij-Katiyar ! Let me know what you think.

File: server/plugin/plugin.go (hunk 3 above)
Comment (Kshitij-Katiyar) [reply server/plugin/plugin.go:371]: sure @javaguirre Actually i was thinking the same but reverted it due to this comment https://github.com/LokeshN/mattermost-plugin-github/pull/1#discussion_r934835632

File: server/plugin/plugin.go (hunk 1 above)
Comment (hanzei) [inline server/plugin/plugin.go:327]: Let's use a proper context with a timeout here, as this method should not run for too long.

File: server/plugin/plugin.go [hunk 4]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
...
+
+	orgRepo, id, objectType, ok := p.getPostPropsForReaction(reaction)
+	if !ok {
+		return
+	}
+
+	info, appErr := p.getGitHubUserInfo(reaction.UserId)
+	if appErr != nil {
+		if appErr.ID != apiErrorIDNotConnected {
+			p.API.LogDebug("Error in getting user info", "error", appErr.Message)
```
Comment (hanzei) [inline server/plugin/plugin.go:322]: Why not `Error()`?
```suggestion
			p.API.LogDebug("Error in getting user info", "error", appErr.Error())
```

File: server/plugin/plugin.go [hunk 5]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
 	return nil
 }
 
+func (p *Plugin) getPostPropsForReaction(reaction *model.Reaction) (orgRepo []string, id float64, objectType string, ok bool) {
+	post, err := p.client.Post.GetPost(reaction.PostId)
+	if err != nil {
+		p.API.LogDebug("Error fetching post for reaction", "error", err.Error())
+		return orgRepo, id, objectType, false
+	}
```
Comment (hanzei) [inline server/plugin/plugin.go:282]: This is costly. For most of the reactions posted, an RPC request is made to fetch the post. It's a bummer `ReactionHasBeenAdded` doesn't provide the post.

Anyway, I don't see a way to get around the RPC call without changing the hook.

File: server/plugin/plugin.go [hunk 6]
```diff
@@ -250,6 +274,151 @@ func (p *Plugin) OnDeactivate() error {
...
+	repo, ok := post.GetProp(postPropGithubRepo).(string)
+	if !ok || repo == "" {
+		return orgRepo, id, objectType, false
+	}
+
+	orgRepo = strings.Split(repo, "/")
+	if len(orgRepo) != 2 {
+		p.API.LogDebug("Invalid organization repository")
+		return orgRepo, id, objectType, false
+	}
```
Comment (hanzei) [inline server/plugin/plugin.go:293]: 0/5 can the caller do the splitting?

Comment (hanzei) [reply server/plugin/plugin.go:293]: Or do some more processing here and return both the org and the repo.

**Review by hanzei — APPROVED** [review]
Nice work :+1:

**Review by javaguirre — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann Gentle reminder to review the PR

**Comment by Kshitij-Katiyar** [comment]
@hanzei synced with master

**Comment by DHaussermann** [comment]
Thanks @Kshitij-Katiyar 
For the supported reactions:
- User reactions in MM will add in GitHub
- User removing from MM will remove the reaction 
- Users not mapped to a GitHub account have no affect (as expected as there would be no user to attribute the reaction to)
- User reacting in GitHub has no effect on MM post (as expected)
- Tested on Issue/PR creations and Issue/PR comments
**Note** I'm having an issue making the `pull_reviews` feature work but it's just something that's not clear about the subscriptions in general. I'm fine confirming this post merge. 

LGTM!

**Comment by mickmister** [comment]
It looks like we don't take into account skin-tone emojis here. Maybe a good HW ticket to support this. More info here https://github.com/LokeshN/mattermost-plugin-github/pull/1#discussion_r934799384

**Pull request merged**